File: CinephoriaServer/Services/Seat/SeatService.cs
using AutoMapper;
using CinephoriaServer.Configurations;
using CinephoriaServer.Models.PostgresqlDb;
using CinephoriaServer.Repository;
using Microsoft.Extensions.Logging;

namespace CinephoriaServer.Services
{
    public class SeatService : ISeatService
    {
        private readonly IUnitOfWorkPostgres _unitOfWork;
        private readonly IMapper _mapper;
        private readonly ILogger<SeatService> _logger;

        public SeatService(IUnitOfWorkPostgres unitOfWork, IMapper mapper, ILogger<SeatService> logger)
        {
            _unitOfWork = unitOfWork;
            _mapper = mapper;
            _logger = logger;
        }

        /// <summary>
        /// Récupère la liste des sièges disponibles pour une séance spécifique.
        /// </summary>
        /// <param name="sessionId">L'identifiant de la séance.</param>
        /// <returns>Une liste de sièges disponibles sous forme de DTO.</returns>
        public async Task<List<SeatDto>> GetAvailableSeatsAsync(int sessionId)
        {
            var seats = await _unitOfWork.Seats.GetAvailableSeatsAsync(sessionId);
            var seatDtos = _mapper.Map<List<SeatDto>>(seats);

            _logger.LogInformation("{Count} sièges disponibles récupérés pour la séance avec l'ID {SessionId}.", seatDtos.Count, sessionId);
            return seatDtos;
        }

        /// <summary>
        /// Ajoute un siège réservé pour les personnes à mobilité réduite dans une salle de cinéma.
        /// </summary>
        /// <param name="theaterId">L'identifiant de la salle de cinéma.</param>
        /// <param name="seatNumber">Le numéro du siège à ajouter.</param>
        /// <returns>Une réponse indiquant le succès ou l'échec de l'opération.</returns>
        public async Task<bool> AddHandicapSeatAsync(int theaterId, string seatNumber)
        {
            await _unitOfWork.Seats.AddHandicapSeatAsync(theaterId, seatNumber);
            await _unitOfWork.CompleteAsync();
            _logger.LogInformation("Siège réservé pour personnes à mobilité réduite ajouté avec succès dans la salle avec l'ID {TheaterId}.", theaterId);
            return true;
        }

        /// <summary>
        /// Supprime un siège réservé pour les personnes à mobilité réduite dans une salle de cinéma.
        /// </summary>
        /// <param name="theaterId">L'identifiant de la salle de cinéma.</param>
        /// <param name="seatNumber">Le numéro du siège à supprimer.</param>
        /// <returns>Une réponse indiquant le succès ou l'échec de l'opération.</returns>
        public async Task<bool> RemoveHandicapSeatAsync(int theaterId, string seatNumber)
        {
            await _unitOfWork.Seats.RemoveHandicapSeatAsync(theaterId, seatNumber);
            await _unitOfWork.CompleteAsync();
            _logger.LogInformation("Siège réservé pour personnes à mobilité réduite supprimé avec succès dans la salle avec l'ID {TheaterId}.", theaterId);
            return true;
        }
    }
}

File: WebApp/src/app/core/services/seat.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { SeatDto, AddHandicapSeatDto, RemoveHandicapSeatDto, UpdateSeatDto, Seat } from '../models/seat.models';
import { environment } from 'src/environments/environment';

@Injectable({
  providedIn: 'root'
})
export class SeatsService {
  private apiUrl = `${environment.apiUrl}/Seats`;

  constructor(private http: HttpClient) {}

  /**
   * Récupère la liste des sièges disponibles pour une séance spécifique.
   * @param sessionId L'identifiant de la séance.
   * @returns Observable contenant une liste de sièges disponibles.
   */
  getAvailableSeats(sessionId: number): Observable<SeatDto[]> {
    return this.http.get<SeatDto[]>(`${this.apiUrl}/available/${sessionId}`);
  }

  /**
   * Ajoute un siège réservé pour les personnes à mobilité réduite dans une salle de cinéma.
   * @param addHandicapSeatDto Les données du siège à ajouter (AddHandicapSeatDto).
   * @returns Observable contenant un message de réussite.
   */
  addHandicapSeat(addHandicapSeatDto: AddHandicapSeatDto): Observable<{ Message: string }> {
    return this.http.post<{ Message: string }>(`${this.apiUrl}/handicap-add-seat`, addHandicapSeatDto);
  }

   /**
   * Récupère tous les sièges d'une salle de cinéma par son identifiant.
   * @param theaterId L'identifiant de la salle.
   * @returns Observable contenant la liste des sièges.
   */
   getSeatsByTheaterId(theaterId: number): Observable<Seat[]> {
    return this.http.get<Seat[]>(`${this.apiUrl}/theater/${theaterId}`);
  }

  /**
   * Met à jour un siège existant.
   * @param updateSeatDto Les nouvelles informations du siège.
   * @returns Observable contenant un message de réussite.
   */
   updateSeat(updateSeatDto: UpdateSeatDto): Observable<{ Message: string }> {
    return this.http.put<{ Message: string }>(`${this.apiUrl}/update`, updateSeatDto);
  }

  /**
   * Supprime un siège réservé pour les personnes à mobilité réduite dans une salle de cinéma.
   * @param dto Les données du siège à supprimer (RemoveHandicapSeatDto).
   * @returns Observable contenant un message de réussite.
   */
  removeHandicapSeat(removeHandicapSeatDto: RemoveHandicapSeatDto): Observable<{ Message: string }> {
    return this.http.delete<{ Message: string }>(`${this.apiUrl}/handicap-delete-seat`, { body: removeHandicapSeatDto });
  }
}